File: attocode/src/providers/resilient-fetch.ts
/**
 * Resilient Fetch Utility
 *
 * Provides network resilience for LLM provider requests:
 * - Configurable timeout to prevent infinite hangs
 * - Retry with exponential backoff for transient failures
 * - Rate limit handling with Retry-After header parsing
 * - Cancellation token support for user interrupts
 */

// =============================================================================
// TYPES
// =============================================================================

/**
 * Network configuration for resilient fetch.
 */
export interface NetworkConfig {
  /** Request timeout in milliseconds (default: 30000) */
  timeout?: number;
  /** Maximum retry attempts (default: 3) */
  maxRetries?: number;
  /** Base delay between retries in ms (default: 1000) */
  baseRetryDelay?: number;
  /** Maximum delay between retries in ms (default: 30000) */
  maxRetryDelay?: number;
  /** HTTP status codes that trigger retry (default: [429, 500, 502, 503, 504]) */
  retryableStatusCodes?: number[];
  /** Extra retry attempts for HTTP 429 (default: 2, so total = maxRetries + 2) */
  maxRetriesFor429?: number;
}

/**
 * Cancellation token interface for external cancellation.
 */
export interface CancellationToken {
  /** Whether cancellation has been requested */
  isCancelled: boolean;
  /** Register a callback to be called on cancellation */
  onCancel?: (callback: () => void) => void;
}

/**
 * Options for resilient fetch.
 */
export interface ResilientFetchOptions {
  /** URL to fetch */
  url: string;
  /** Fetch init options (method, headers, body, etc.) */
  init: RequestInit;
  /** Provider name for error messages */
  providerName: string;
  /** Network configuration */
  networkConfig?: NetworkConfig;
  /** External cancellation token */
  cancellationToken?: CancellationToken;
  /** Callback for retry events */
  onRetry?: (attempt: number, delay: number, error: Error) => void;
}

/**
 * Result of resilient fetch.
 */
export interface ResilientFetchResult {
  /** The fetch response */
  response: Response;
  /** Number of attempts made */
  attempts: number;
  /** Total duration in milliseconds */
  duration: number;
}

/**
 * Error thrown when fetch fails after all retries.
 */
export class ResilientFetchError extends Error {
  constructor(
    message: string,
    public readonly providerName: string,
    public readonly attempts: number,
    public readonly lastError?: Error,
    public readonly isTimeout: boolean = false,
    public readonly isCancelled: boolean = false
  ) {
    super(message);
    this.name = 'ResilientFetchError';
  }
}

// =============================================================================
// DEFAULT CONFIG
// =============================================================================

const DEFAULT_CONFIG: Required<NetworkConfig> = {
  timeout: 30000,
  maxRetries: 3,
  baseRetryDelay: 1000,
  maxRetryDelay: 30000,
  retryableStatusCodes: [429, 500, 502, 503, 504],
  maxRetriesFor429: 2,
};

// =============================================================================
// RESILIENT FETCH
// =============================================================================

/**
 * Perform a fetch with timeout, retry, and cancellation support.
 *
 * @example
 * ```typescript
 * const { response, attempts, duration } = await resilientFetch({
 *   url: 'https://api.anthropic.com/v1/messages',
 *   init: { method: 'POST', headers: {...}, body: JSON.stringify(body) },
 *   providerName: 'anthropic',
 *   networkConfig: { timeout: 60000, maxRetries: 3 },
 * });
 * ```
 */
export async function resilientFetch(
  options: ResilientFetchOptions
): Promise<ResilientFetchResult> {
  const {
    url,
    init,
    providerName,
    networkConfig = {},
    cancellationToken,
    onRetry,
  } = options;

  const config = { ...DEFAULT_CONFIG, ...networkConfig };
  const startTime = Date.now();
  let lastError: Error | undefined;
  let attempts = 0;

  // Use max possible retries (429 may get extra attempts)
  const maxPossibleRetries = config.maxRetries + config.maxRetriesFor429;
  while (attempts < maxPossibleRetries) {
    attempts++;

    // Check for cancellation before each attempt
    if (cancellationToken?.isCancelled) {
      throw new ResilientFetchError(
        `Request cancelled by user`,
        providerName,
        attempts,
        lastError,
        false,
        true
      );
    }

    // Create AbortController for timeout
    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), config.timeout);

    // Link external cancellation token
    let cancelCleanup: (() => void) | undefined;
    if (cancellationToken?.onCancel) {
      cancellationToken.onCancel(() => controller.abort());
      // Note: In a real implementation, you'd want proper cleanup here
    }

    try {
      const response = await fetch(url, {
        ...init,
        signal: controller.signal,
      });

      clearTimeout(timeoutId);
      if (cancelCleanup) cancelCleanup();

      // Check if response is retryable
      if (config.retryableStatusCodes.includes(response.status)) {
        const is429 = response.status === 429;
        // 429 gets extra retry budget
        const effectiveMaxRetries = is429
          ? config.maxRetries + config.maxRetriesFor429
          : config.maxRetries;

        // Parse Retry-After header for rate limits
        const retryAfter = parseRetryAfter(response.headers.get('Retry-After'));
        // 429 uses steeper backoff (3^n instead of 2^n)
        const delay = retryAfter ?? (is429
          ? calculate429Backoff(attempts, config)
          : calculateBackoff(attempts, config));

        lastError = new Error(`HTTP ${response.status}: ${response.statusText}`);

        // Don't retry if this is the last attempt
        if (attempts >= effectiveMaxRetries) {
          throw new ResilientFetchError(
            `${providerName} request failed after ${attempts} attempts: HTTP ${response.status}`,
            providerName,
            attempts,
            lastError
          );
        }

        // Notify about retry
        if (onRetry) {
          onRetry(attempts, delay, lastError);
        }

        await sleep(delay);
        continue;
      }

      // Success (or non-retryable error - let caller handle)
      return {
        response,
        attempts,
        duration: Date.now() - startTime,
      };
    } catch (error) {
      clearTimeout(timeoutId);
      if (cancelCleanup) cancelCleanup();

      // Handle abort (timeout or cancellation)
      if (error instanceof Error && error.name === 'AbortError') {
        if (cancellationToken?.isCancelled) {
          throw new ResilientFetchError(
            `Request cancelled by user`,
            providerName,
            attempts,
            error,
            false,
            true
          );
        }
        // Timeout
        lastError = new Error(`Request timeout after ${config.timeout}ms`);

        if (attempts >= config.maxRetries) {
          throw new ResilientFetchError(
            `${providerName} request timed out after ${attempts} attempts`,
            providerName,
            attempts,
            lastError,
            true
          );
        }

        const delay = calculateBackoff(attempts, config);
        if (onRetry) {
          onRetry(attempts, delay, lastError);
        }
        await sleep(delay);
        continue;
      }

      // Network error (ECONNREFUSED, DNS failure, etc.)
      lastError = error instanceof Error ? error : new Error(String(error));

      if (attempts >= config.maxRetries) {
        throw new ResilientFetchError(
          `${providerName} network error after ${attempts} attempts: ${lastError.message}`,
          providerName,
          attempts,
          lastError
        );
      }

      const delay = calculateBackoff(attempts, config);
      if (onRetry) {
        onRetry(attempts, delay, lastError);
      }
      await sleep(delay);
    }
  }

  // Should not reach here, but just in case
  throw new ResilientFetchError(
    `${providerName} request failed after ${attempts} attempts`,
    providerName,
    attempts,
    lastError
  );
}

// =============================================================================
// HELPERS
// =============================================================================

/**
 * Parse Retry-After header value.
 * Can be either a number of seconds or an HTTP-date.
 */
function parseRetryAfter(header: string | null): number | null {
  if (!header) return null;

  // Try parsing as number of seconds
  const seconds = parseInt(header, 10);
  if (!isNaN(seconds)) {
    return seconds * 1000; // Convert to milliseconds
  }

  // Try parsing as HTTP-date
  try {
    const date = new Date(header);
    const delay = date.getTime() - Date.now();
    return delay > 0 ? delay : null;
  } catch {
    return null;
  }
}

/**
 * Calculate exponential backoff delay with jitter.
 */
function calculateBackoff(
  attempt: number,
  config: Required<NetworkConfig>
): number {
  // Exponential backoff: baseDelay * 2^(attempt-1)
  const exponentialDelay = config.baseRetryDelay * Math.pow(2, attempt - 1);

  // Add jitter (±25%)
  const jitter = exponentialDelay * 0.25 * (Math.random() * 2 - 1);
  const delay = exponentialDelay + jitter;

  // Clamp to max delay
  return Math.min(delay, config.maxRetryDelay);
}

/**
 * Steeper exponential backoff for 429 rate limit errors.
 * Uses 3^n instead of 2^n to give rate limit windows more recovery time.
 */
function calculate429Backoff(
  attempt: number,
  config: Required<NetworkConfig>
): number {
  // Steeper backoff: baseDelay * 3^(attempt-1)
  const exponentialDelay = config.baseRetryDelay * Math.pow(3, attempt - 1);

  // Add jitter (±25%)
  const jitter = exponentialDelay * 0.25 * (Math.random() * 2 - 1);
  const delay = exponentialDelay + jitter;

  // Clamp to max delay
  return Math.min(delay, config.maxRetryDelay);
}

/**
 * Sleep for a given number of milliseconds.
 */
function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

// =============================================================================
// UTILITIES
// =============================================================================

/**
 * Check if an error is a resilient fetch error.
 */
export function isResilientFetchError(error: unknown): error is ResilientFetchError {
  return error instanceof ResilientFetchError;
}

/**
 * Check if an error is a timeout error.
 */
export function isTimeoutError(error: unknown): boolean {
  return isResilientFetchError(error) && error.isTimeout;
}

/**
 * Check if an error is a cancellation error.
 */
export function isCancellationError(error: unknown): boolean {
  return isResilientFetchError(error) && error.isCancelled;
}

File: src/attocode/integrations/utilities/resilient_fetch.py
"""Resilient HTTP fetch with retry, timeout, and circuit breaker.

Wraps httpx with retry logic, exponential backoff, rate limiting,
and circuit breaker pattern for reliable HTTP requests to LLM APIs.
"""

from __future__ import annotations

import asyncio
import time
from dataclasses import dataclass, field
from typing import Any


@dataclass(slots=True)
class FetchResult:
    """Result of a resilient fetch operation."""

    status_code: int
    body: bytes
    headers: dict[str, str]
    success: bool
    error: str = ""
    retries: int = 0
    duration_ms: float = 0.0
    provider_name: str = ""

    @property
    def text(self) -> str:
        return self.body.decode("utf-8", errors="replace")

    @property
    def json(self) -> Any:
        import json
        return json.loads(self.body)


@dataclass(slots=True)
class FetchConfig:
    """Configuration for resilient fetch."""

    timeout: float = 120.0
    max_retries: int = 3
    retry_base_delay: float = 1.0
    retry_max_delay: float = 30.0
    retry_on_status: frozenset[int] = frozenset({429, 500, 502, 503, 504})
    rate_limit_delay: float = 0.0  # Extra delay between requests (0 = disabled)


class CircuitBreakerState:
    """Simple circuit breaker for a provider."""

    def __init__(self, threshold: int = 5, timeout: float = 60.0) -> None:
        self.threshold = threshold
        self.timeout = timeout
        self.failure_count = 0
        self.last_failure_time = 0.0
        self.is_open = False

    def record_success(self) -> None:
        self.failure_count = 0
        self.is_open = False

    def record_failure(self) -> None:
        self.failure_count += 1
        self.last_failure_time = time.time()
        if self.failure_count >= self.threshold:
            self.is_open = True

    def should_allow(self) -> bool:
        if not self.is_open:
            return True
        # Half-open: allow after timeout period
        if time.time() - self.last_failure_time > self.timeout:
            return True
        return False


class ResilientFetch:
    """HTTP client with retry, timeout, and circuit breaker.

    Wraps httpx.AsyncClient with resilience patterns for
    reliable communication with LLM provider APIs.
    """

    def __init__(self, config: FetchConfig | None = None) -> None:
        self._config = config or FetchConfig()
        self._circuit_breakers: dict[str, CircuitBreakerState] = {}
        self._client: Any = None  # Lazy httpx.AsyncClient

    async def _get_client(self) -> Any:
        """Get or create the httpx client."""
        if self._client is None:
            import httpx
            self._client = httpx.AsyncClient(timeout=self._config.timeout)
        return self._client

    async def close(self) -> None:
        """Close the HTTP client."""
        if self._client is not None:
            await self._client.aclose()
            self._client = None

    def _get_circuit_breaker(self, provider: str) -> CircuitBreakerState:
        """Get or create a circuit breaker for a provider."""
        if provider not in self._circuit_breakers:
            self._circuit_breakers[provider] = CircuitBreakerState()
        return self._circuit_breakers[provider]

    async def fetch(
        self,
        method: str,
        url: str,
        *,
        headers: dict[str, str] | None = None,
        body: bytes | str | None = None,
        json_body: Any = None,
        provider: str = "",
    ) -> FetchResult:
        """Perform a resilient HTTP fetch.

        Args:
            method: HTTP method (GET, POST, etc.).
            url: Request URL.
            headers: Request headers.
            body: Raw body bytes or string.
            json_body: JSON body (mutually exclusive with body).
            provider: Provider name for circuit breaker tracking.

        Returns:
            FetchResult with response data.
        """
        # Circuit breaker check
        if provider:
            cb = self._get_circuit_breaker(provider)
            if not cb.should_allow():
                return FetchResult(
                    status_code=0,
                    body=b"",
                    headers={},
                    success=False,
                    error=f"Circuit breaker open for provider '{provider}'",
                    provider_name=provider,
                )

        client = await self._get_client()
        last_error = ""
        retries = 0

        for attempt in range(self._config.max_retries + 1):
            start = time.time()
            try:
                # Rate limiting delay
                if self._config.rate_limit_delay > 0 and attempt > 0:
                    await asyncio.sleep(self._config.rate_limit_delay)

                # Build request kwargs
                kwargs: dict[str, Any] = {"headers": headers or {}}
                if json_body is not None:
                    kwargs["json"] = json_body
                elif body is not None:
                    kwargs["content"] = body

                response = await client.request(method, url, **kwargs)
                duration_ms = (time.time() - start) * 1000

                # Success
                if response.status_code < 400:
                    if provider:
                        self._get_circuit_breaker(provider).record_success()
                    return FetchResult(
                        status_code=response.status_code,
                        body=response.content,
                        headers=dict(response.headers),
                        success=True,
                        retries=retries,
                        duration_ms=duration_ms,
                        provider_name=provider,
                    )

                # Check if we should retry this status code
                if response.status_code in self._config.retry_on_status:
                    last_error = f"HTTP {response.status_code}"
                    retries += 1

                    # Rate limit: check Retry-After header
                    retry_after = response.headers.get("retry-after", "")
                    if retry_after:
                        try:
                            delay = float(retry_after)
                        except ValueError:
                            delay = self._calculate_backoff(attempt)
                    else:
                        delay = self._calculate_backoff(attempt)

                    await asyncio.sleep(delay)
                    continue

                # Non-retryable error
                if provider:
                    self._get_circuit_breaker(provider).record_failure()

                return FetchResult(
                    status_code=response.status_code,
                    body=response.content,
                    headers=dict(response.headers),
                    success=False,
                    error=f"HTTP {response.status_code}",
                    retries=retries,
                    duration_ms=duration_ms,
                    provider_name=provider,
                )

            except Exception as e:
                duration_ms = (time.time() - start) * 1000
                last_error = str(e)
                retries += 1

                if attempt < self._config.max_retries:
                    delay = self._calculate_backoff(attempt)
                    await asyncio.sleep(delay)
                    continue

                if provider:
                    self._get_circuit_breaker(provider).record_failure()

                return FetchResult(
                    status_code=0,
                    body=b"",
                    headers={},
                    success=False,
                    error=last_error,
                    retries=retries,
                    duration_ms=duration_ms,
                    provider_name=provider,
                )

        # Should not reach here, but just in case
        return FetchResult(
            status_code=0,
            body=b"",
            headers={},
            success=False,
            error=f"Max retries exceeded: {last_error}",
            retries=retries,
            provider_name=provider,
        )

    def _calculate_backoff(self, attempt: int) -> float:
        """Calculate exponential backoff delay."""
        delay = self._config.retry_base_delay * (2 ** attempt)
        return min(delay, self._config.retry_max_delay)
